feat: Implementation of resample_spatial to reproject values

SerRichard opened the pull request:
Dask implementation of resample spatial using the odc.algo reproject function. I've tried to handle some trickier parts (like infering the affine transformation) with some helper utils. Ready for review when all boxes are ticked!

- [X] Reprojection
- [x] Resampling
- [x] Unit tests

LukeWeidenwalker commented:
Just to leave it here too: Since https://github.com/opendatacube/odc-geo/pull/88 came out, it would probably be good to see whether this already does this!

SerRichard commented:
> Just to leave it here too: Since [opendatacube/odc-geo#88](https://github.com/opendatacube/odc-geo/pull/88) came out, it would probably be good to see whether this already does this!

Switched implementation to use the odc.reproject function, much less code to maintain now!

@ValentinaHutter One thing I'm not sure on. Odc reproject will automatically resample the raster to the provided projection. In the case a user provides no resolution to resample, and only wants a reproject, should we force the resolution of the input cube? Or just use accept resolution as none? Not sure what the preference would be.

SerRichard commented:
@ValentinaHutter One other concern, we've no way or reliably enforcing the convention of the user specifying the target resolution in the units of the output crs. Scao recommended this could be enforced with accepted rangers for different CRsystems. Example, if the user incorrectly specifies 0.0009 as their output resolution, and the crs is in metres, this will run for longer. Not sure how much longer on dask, so may not be a huge concern.

ValentinaHutter reviewed (commented):
LGTM!

ValentinaHutter commented:
> > Just to leave it here too: Since [opendatacube/odc-geo#88](https://github.com/opendatacube/odc-geo/pull/88) came out, it would probably be good to see whether this already does this!
> 
> Switched implementation to use the odc.reproject function, much less code to maintain now!
> 
> @ValentinaHutter One thing I'm not sure on. Odc reproject will automatically resample the raster to the provided projection. In the case a user provides no resolution to resample, and only wants a reproject, should we force the resolution of the input cube? Or just use accept resolution as none? Not sure what the preference would be.

In https://processes.openeo.org/#resample_spatial the default resolution=0 and says that it `Doesn't change the resolution by default (0).` So, I think this means, we need to keep the input resolution?

ValentinaHutter commented:
> @ValentinaHutter One other concern, we've no way or reliably enforcing the convention of the user specifying the target resolution in the units of the output crs. Scao recommended this could be enforced with accepted rangers for different CRsystems. Example, if the user incorrectly specifies 0.0009 as their output resolution, and the crs is in metres, this will run for longer. Not sure how much longer on dask, so may not be a huge concern.

Hm, it sounds quite complicated to check this with different ranges for every crs, what do you think? Also, the definition says, that the resolution is `Specified in the units of the target projection.` So, the user should be aware of this, and we can probably assume, that it is in the required unit. Does `data.odc.reproject` also assume, that the unit of the resolution matches the output crs?

SerRichard commented on openeo_processes_dask/process_implementations/cubes/resample.py:
```diff
@@ -0,0 +1,73 @@
+    projection: Union[str, int] = None,
+    resolution: Union[str, int] = None,
+    method: str = "near",
+    align: str = "upper-left",
```
Just logged a warning with the use of the function

SerRichard commented on tests/test_resample.py:
```diff
@@ -0,0 +1,53 @@
+    assert output_cube.odc.spatial_dims == ("y", "x")
+    assert output_cube.rio.crs == CRS.from_user_input(output_crs)
+
+    if output_crs != "4326":
```
added :)

SerRichard commented on openeo_processes_dask/process_implementations/cubes/resample.py:
```diff
@@ -0,0 +1,73 @@
+    reprojected.attrs["crs"] = data.rio.crs
+
+    # Undo odc specific re-ordering.
+    reprojected = reprojected.transpose("bands", "t", "x", "y")
```
added :)

LukeWeidenwalker commented on openeo_processes_dask/process_implementations/cubes/resample.py:
```diff
@@ -0,0 +1,79 @@
+def resample_spatial(
+    data: RasterCube,
+    projection: Union[str, int] = None,
+    resolution: Union[str, int] = None,
```
According to the spec this wants to be 0 by default!

SerRichard replied:
Updated!

LukeWeidenwalker commented on openeo_processes_dask/process_implementations/cubes/resample.py:
```diff
@@ -0,0 +1,79 @@
+        )
+
+    # Re-order, this is specifically done for odc reproject
+    data_cp = data.transpose("bands", "t", "y", "x")
```
This will fail if the dimensions are named differently - think you can use the `.openeo` accessor to find the names dynamically instead!

LukeWeidenwalker commented on openeo_processes_dask/process_implementations/cubes/resample.py:
```diff
@@ -0,0 +1,79 @@
+        projection = data_cp.rio.crs
+
+    try:
+        projection = CRS(projection)
```
Maybe initialise this with `CRS.from_user_input`

SerRichard replied:
This function just wraps this call, so not sure of the benefit?

SerRichard replied:
eh changed ;)

LukeWeidenwalker commented on openeo_processes_dask/process_implementations/cubes/resample.py:
```diff
@@ -0,0 +1,79 @@
+):
+    """Resamples the spatial dimensions (x,y) of the data cube to a specified resolution and/or warps the data cube to the target projection. At least resolution or projection must be specified."""
+
+    if align == "upper-left":
```
1) This should especially appear if the user chose a different option than the default
2) If we don't support it, shouldn't we just remove it from the spec we expose?

SerRichard replied:
Yeah happy to remove it!

LukeWeidenwalker commented on openeo_processes_dask/process_implementations/cubes/resample.py:
```diff
@@ -0,0 +1,79 @@
+def resample_spatial(
+    data: RasterCube,
+    projection: Union[str, int] = None,
+    resolution: Union[str, int] = None,
```
Why can this be a string?

clausmichele replied:
The projection can be a string if provided as WKT or PROJ string. If it's an int it should be an EPSG code.
The resolution maybe should be int and float instead of string?

SerRichard replied:
yeah my mistake, the resolution should be an int!

LukeWeidenwalker reviewed (changes requested):
Sorry for the delay on this review - nice use of odc.geo!
Couple of comments!

LukeWeidenwalker commented:
Notes from our debugging session today:

- We observed the following to happen with a reprojection to EQUI7: 
![image](https://github.com/Open-EO/openeo-processes-dask/assets/17790923/c987e780-d305-4e4f-924a-1fe49a41ff0c)
 It should be possible to quick fix this by reprojecting the bounding box and cropping. This seems to be an issue solely for EQUI7, we've tried reprojecting to a few other projections and didn't encounter the same problem.
- The transpose does not work if the dimension names have changed. We don't think this transpose is really required, so can just get rid of this

SerRichard commented:
Have some time today, so will pick this back up and try to resolve the retrojection with equi7 issue we discovered last month + resolve test issues

SerRichard commented:
Using odc.geo 0.4.1, I was unable to reproduce what we saw last month one both fake and real (sen2like output) data. Fixed odc minimum version to 0.4.1, fixed tests and accessor import for odc, and reviewed output arrays with @LukeWeidenwalker!

LukeWeidenwalker commented on openeo_processes_dask/process_implementations/cubes/resample.py:
```diff
@@ -0,0 +1,77 @@
+
+def resample_spatial(
+    data: RasterCube,
+    projection: Union[str, int] = None,
```
```suggestion
    projection: Optional[Union[str, int]] = None,
```

LukeWeidenwalker commented on openeo_processes_dask/process_implementations/cubes/resample.py:
```diff
@@ -0,0 +1,77 @@
+        data.openeo.x_dim,
+    )
+
+    if not projection:
```
```suggestion
    if projection is None:
```

LukeWeidenwalker commented on openeo_processes_dask/process_implementations/cubes/resample.py:
```diff
@@ -0,0 +1,77 @@
+    except CRSError as e:
+        raise CRSError(f"{projection} Can not be parsed to CRS.")
+
+    if not resolution:
```
```suggestion
    if resolution == 0:
```

Just to be explicit!

LukeWeidenwalker commented on openeo_processes_dask/process_implementations/cubes/resample.py:
```diff
@@ -0,0 +1,77 @@
+    try:
+        projection = CRS.from_user_input(projection)
+    except CRSError as e:
+        raise CRSError(f"{projection} Can not be parsed to CRS.")
```
```suggestion
        raise CRSError(f"Provided projection string: '{projection}' can not be parsed to CRS.") from e
```

Would include the original traceback here

LukeWeidenwalker commented on openeo_processes_dask/process_implementations/cubes/resample.py:
```diff
@@ -0,0 +1,77 @@
+        how=projection, resolution=resolution, resampling=method
+    )
+
+    if "longitude" in reprojected.dims:
```
I think this renaming should only happen if the original dimension name was x/y too.

LukeWeidenwalker reviewed (changes requested):
Looking good! Couple of minor comments, but once these are resolved we should be able to finally merge this!

LukeWeidenwalker commented on openeo_processes_dask/process_implementations/cubes/utils.py:
```diff
@@ -1,8 +1,8 @@
-
 except ImportError:
     dask = None
+import math
```
This looks unintentional :D

LukeWeidenwalker commented:
Ahh, two more things:
- the spec allows the `align` parameter, we should at least accept that and log a warning that it doesn't do anything currently
- the spec needs to be updated in openeo-processes

SerRichard commented:
spec added in eodcgmbh/openeo-processes

clausmichele commented:
@SerRichard I have a question: after putting so much work into this process implementation, did you ever test it with an openEO process graph? Because currently it is not imported here https://github.com/Open-EO/openeo-processes-dask/blob/main/openeo_processes_dask/process_implementations/cubes/__init__.py and therefore can't be used.

clausmichele commented:
Additionally, not exposing a parameter like `align` just because it's not supported it's a bad idea in my opinion.
First of all since it's a parameter with a default value, which could be added by default added to the process graph by the clients (Python client adds it).

@soxofaan does the Python client always add to the process graph all the optional parameters with default values?

soxofaan commented:
> @soxofaan does the Python client always add to the process graph all the optional parameters with default values?

no we don't always automatically add the default value of unset parameters, it depends a bit on details of `DataCube` method.

In case of `resample_spatial` however, we do always add the `align` parameter with default value "upper-left"

m-mohr commented:
> Additionally, not exposing a parameter like align just because it's not supported it's a bad idea in my opinion.

Actually, it's the opposite. It's meant to be that if something is not supported, it's not exposed. Process definitions should reflect the implementation status.

> In case of resample_spatial however, we do always add the align parameter with default value "upper-left"

I think this should be changed. It should only be sent if not the default value and supported by the back-end.

clausmichele commented:
Thanks @m-mohr, so as you said, the Python Client needs to be adjusted.

soxofaan commented:
-> https://github.com/Open-EO/openeo-python-client/issues/500

m-mohr commented:
Yes, you are right. If one value is supported, the backend should change the schema accordingly so that it requires exactly that value. If a parameter is not supported at all, it should be removed.

That still would require some kind of change in the Python client, I think.
